### DISCUSSION TITLE
Generalized Tokenization Rules

### DISCUSSION THREAD

**Discussion by marshallward** [marshallward/flint#14] [General]
This discussion is for consolidation of rules for the Fortran lexer, and the general rules for efficiently constructing the tokens, independent of language.

Pulling from #11, this is @certik's suggested procedure:

> A robust Fortran parser thus might consists of:
> 
> 1. Pre-tokenizing phase: take care of &, perhaps other things
> 2. Tokenizing: determine token types based on regex (or equivalent such as re2c or hand tokenizer)
> 3. Pre-parsing: determine with certainty the token types by keeping track of the context and looking ahead
> 4. Parsing: Can assume the token types are correct

Flint's approach is similar, but not identical and with its own particular flaws.

1. Generic tokenization:  This is a low-level operation which distinguishes between names, symbols, and whitespace.  Tokens are returned as a list of lists, where each list corresponds to a line of the source file.
   
   * This is where the difficult problem of string delimiters, escaped characters, and comments is handled.
   * Numerical tokens are fully constructed here, particularly the somewhat tricky floating point format.
   * It does not attempt to classify the symbols.
   * There is also no "bundling" of tokens, such as the index parentheses of an array.  They are stored serially.
   * Notably, it also makes no effort to handle line continuations (`&`), which is handled in the second step.  This is a mistake, as explained below, and would be better handled here.
   * **Edit**: "Null" tokens like whitespace are resolved here, but are also discarded in the final product.
   
2. Statement construction:  This is an iterator (named `FortLines`) which pulls from the list of tokens and returns the next statement.

   * Despite the elegant description, the main task of `FortLines` is to handle the line continuations (`&`) and compound statements (`;`).  This iterator will cache, store, and keep track of what it needs to return the next logical statement from the jumble of tokens.
   * **Edit**: This is also where empty lines (and comment lines) are discarded.
   
3. Parser: This is where we try to actually identify the tokens and start building up the actual program structure.

   * This is the part that I generally describe as "ad hoc", since it mostly passes through scattered blocks of Python code which don't follow much general structure, as opposed to a well-documented BNF grammar as might be passed to yacc or bison.
   * Nor is there any formal attempt to build something like an AST, although there is a kind of decision tree and some coherence to the data types which resemble an AST.
   * It is also the least complete, and focuses more on the static content (modules, programs, variable, functions, etc) rather than the expressions and their behavior.

Also, since `FortLines` is an iterator, there is more of a cycle between steps 2 and 3, rather than a complete generation of statements following by a complete parsing.

----

A defining aspect of the flint approach is to delay token classification for as long as possible.  This simplifies the task of building the tokens and constructing the statement, but we do end repeating repeating the iteration over tokens, and perhaps repeating some of the analysis.

We probably still want to do things in stages, but perhaps think about what is learned at each step.

Also it may be beneficial to do the entire thing as an iterator: construct and fetch tokens until you have a complete statement, then return the statement to the parser, before moving onto the next.  This may also be more favorable to the REPL.

**Comment by certik** [comment]
If you are interested in designing this in a way that it could be used by LFortran, let me know. If not, that is fine too, I am fine either way. I am simply trying to find ways to collaborate.

**Reply by marshallward** [reply]
I'd also like to see this happen, even if I can't yet imagine how it would look.

Our discussions of a "roundtrip" parser taught me a great deal on what can and ought to be handled during these tokenization steps.  For example, appending the "null" whitespace and comment tokens is proving much more effective than I would have previously though, and I think it could be extended to include line continuations or empty blank statements, which would have very major impacts on the rest of the lexer.

I think that producing a roundtrip in Python is going to be very useful in demonstrating how a more generalized and powerful lexer should be designed.

But as we discussed, it's always going to be far too slow to be used in production.  Are you thinking of moving it to a general library (say in C), and then creating the hooks to LFortran and Python?

**Reply by certik** [reply]
I would first figure out how to do it, then do it in C in re2c and some pre and post processing. It will be one ore two files. Then we can include them in flint and lfortran.
